src/Processors/Formats/RowInputFormatWithNamesAndTypes.cpp

```diff
@@ -251,6 +252,17 @@ bool RowInputFormatWithNamesAndTypes::readRow(MutableColumns & columns, RowReadE
 
     format_reader->skipRowEndDelimiter();
 
+    if (format_settings.force_null_for_omitted_fields)
+    {
+        for (auto index : column_mapping->not_presented_columns)
+            if (!isNullableOrLowCardinalityNullable(data_types[index]))
+                throw Exception(
+                    ErrorCodes::TYPE_MISMATCH,
+                    "Cannot insert NULL value into a column type '{}' at index {}",
+                    data_types[index]->getName(),
+                    index);
+    }
+
```

Comment:
I think it is better not to do this check each time a row is read, but only once after the header known, because the same values will be present in each row (in other words `not_presented_columns` is populated when the header is read, not when the individiual rows).

Reply:
nice! moved it to the prefix parser.
also added support for `input_format_[c|t]sv_allow_variable_number_of_columns` as @alexey-milovidov requested